hazelcast/src/main/java/com/hazelcast/security/impl/function/SecuredFunction.java

```diff
@@ -34,10 +34,13 @@ public interface SecuredFunction {
 
     /**
      * @return the list of permissions required to run this function
+     *
+     * @implNote If used, permissions should be hardcoded based on function arguments/context.
+     * It is not secure to keep {@code List<Permission>} in the function itself because
+     * it could be tampered with if the function is serialized and passed from untrusted caller.
      */
     @Nullable
     default List<Permission> permissions() {
         return null;
```

Comment:
Could you add a test checking we know about all implementations, that return the `null`?
I mean something similar to what the `MessageTaskSecurityTest` does. (BTW, @JackPGreen is currently working on improvements in message task permissions)
We would have the whitelisted implementations in one place.

Reply:
It is easy to find all classes (also anonymous) implementing `SecuredFunction` eg. in ArchUnit rule, but I did not find any way to find also all lambdas for which no class is generated. So it would only be a partial check, and the easiest way to forget about permissions is by using lambda. But still, maybe such check would be better than nothing.

Reply:
Another idea is to throw exception in default `permissions()` implementation. That way, each function that is used in security-sensitive context (i.e. someone checks permissions for it) will have to explicitly deal with permissions even if it does not need any (in such case it will explicitly return null). But functions that are not used in such contexts will not have to be bothered.

Reply:
I tried the above idea but it has a few problems:

1. you cannot use lambdas -> more code
2. anonymous class defined in instance method always captures `this` unlike lambda. There are places were we create FunctionEx in instance methods. The class would have to be not anonymous.
3. many places to add permissions, we might have limited coverage of uses and compositions, so it possible that we might miss some cases where it is necessary `permissions()`.

Reply:
you can check how it looks like in https://github.com/k-jamroz/hazelcast/tree/functionex-composition-security-default-throw (https://github.com/k-jamroz/hazelcast/commit/a352b75a8e537ffaef088360e5f53ad0729c2bfb) - still not complete.